src/python/pants/engine/rules.py

```diff
@@ -191,13 +192,21 @@ def rule_decorator(func, **kwargs) -> Callable:
         name=f"{func_id} return",
         raise_type=MissingReturnTypeAnnotation,
     )
+
+    func_params = inspect.signature(func).parameters
+    for parameter in param_type_overrides:
+        if parameter not in func_params:
+            raise ValueError(f"Unknown parameter name in `param_type_overrides`: {parameter}")
+
     parameter_types = tuple(
-        _ensure_type_annotation(
+        param_type_overrides[parameter]
+        if parameter in param_type_overrides
+        else _ensure_type_annotation(
```

Comment:
nit: can use `param_type_overrides.get(parameter, fallback)`

Reply:
That requires us to evaluate the fallback argument, which I don't want to do